addon/instance-initializers/prefetch.js

```diff
@@ -38,7 +38,13 @@ export function initialize(instance) {
       }
 
       // Run the prefetch hook if the route has one.
-      handlerInfo.handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
+      if (!handlerInfo.handler && handlerInfo.handlerPromise) {
+        handlerInfo.handlerPromise.then((handler) => {
+          handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
+        });
+      } else {
+        handlerInfo.handler._prefetched = handlerInfo.runSharedModelHook(transition, 'prefetch', [fullParams]);
```

Comment:
Will `handlerPromise` always be available on `handlerInfo`? If it is, I think it would be okay to always promise chain from it. That would DRY this up a bit.